### DISCUSSION TITLE
STM32H745: unable to open 2 sockets at the same time

### DISCUSSION THREAD

**Discussion by joelguittet** [zephyrproject-rtos/zephyr#64875] [Q&A] [answered]
Hello community !

I'm writing an application that need to open multiple sockets at the same time. I'm using `nucleo_h745zi_q_m7` evaluation board with Ethernet interface.

One sockets is fine (http client connecting to a server using TLS, port 443), but then I'm not able to open a new socket. I need to close the first one and wait a bit to permit a second socket to be used.

I don't know at all which settings I should modify, I have already checked at the menuconfig but don't find anything related to this in the Networking or STM32 Ethernet Driver sections.

Thanks for any ideas that may help!

Joel

**Comment by jukkar** [comment]
Please check the values of following options
```
CONFIG_NET_MAX_CONTEXTS
CONFIG_NET_MAX_CONN
CONFIG_POSIX_MAX_FDS
CONFIG_NET_SOCKETS_POLL_MAX
```
The issue is probably related to one of those options to be too small. Note that one TLS connection uses two sockets so you might hit a limit. I suspect the `CONFIG_NET_SOCKETS_POLL_MAX` as I saw similar issue yesterday (default value is 3).

**Comment by joelguittet** [comment]
Thanks @jukkar for the quick answer to this topic!

All these values were the default, so:
- I have increased `CONFIG_NET_SOCKETS_POLL_MAX` to 10 => No changes
- I have increased all of them x4 => No changes

Just to give more details about this, I added an errno log, and the exact error is `socket()` returns -1 with errno 12 (no memory). So I guess the idea is not bad but probably another settings should be changed.

Joel

**Comment by joelguittet** [comment]
Following reading https://github.com/zephyrproject-rtos/zephyr/issues/38191 I have also increased the followings:

CONFIG_NET_BUF_RX_COUNT=64
CONFIG_NET_BUF_TX_COUNT=64
CONFIG_NET_PKT_RX_COUNT=32
CONFIG_NET_PKT_TX_COUNT=32

But still no progress so far :-(

**Comment by jukkar** [comment]
Hmm, perhaps this is related to TLS. Can you try without TLS, do you get the same result?

**Reply by joelguittet** [reply]
Except making a specific application, I'm not able to do it, the distant server is only accepting HTTPS.

**Comment by jukkar** [comment]
The `socket()` returning `ENOMEM` indicates that some limit is too small. And for TLS sockets it might be caused some mbedtls config option and there are many of those.

**Reply by joelguittet** [reply]
Yeah I know particularly about the heap size, configured 60000 Bytes (I hope it's enough!). I have set the following to be exact:

```
CONFIG_MBEDTLS=y
CONFIG_MBEDTLS_BUILTIN=y
CONFIG_MBEDTLS_ENABLE_HEAP=y
CONFIG_MBEDTLS_HEAP_SIZE=60000
CONFIG_MBEDTLS_CFG_FILE="config-tls-generic.h"
CONFIG_MBEDTLS_ECDH_C=y
CONFIG_MBEDTLS_ECDSA_C=y
CONFIG_MBEDTLS_ECP_C=y
CONFIG_MBEDTLS_ECP_DP_SECP256R1_ENABLED=y
CONFIG_MBEDTLS_GENPRIME_ENABLED=y
CONFIG_MBEDTLS_CIPHER_AES_ENABLED=y
CONFIG_MBEDTLS_CIPHER_CCM_ENABLED=y
CONFIG_MBEDTLS_CIPHER_GCM_ENABLED=y
CONFIG_MBEDTLS_KEY_EXCHANGE_ECDHE_PSK_ENABLED=y
CONFIG_MBEDTLS_KEY_EXCHANGE_ECDHE_RSA_ENABLED=y
CONFIG_MBEDTLS_KEY_EXCHANGE_ECDHE_ECDSA_ENABLED=y
CONFIG_MBEDTLS_KEY_EXCHANGE_RSA_ENABLED=y
CONFIG_MBEDTLS_PEM_CERTIFICATE_FORMAT=y
CONFIG_MBEDTLS_PK_WRITE_C=y
CONFIG_MBEDTLS_ZEPHYR_ENTROPY=y
CONFIG_MBEDTLS_SSL_MAX_CONTENT_LEN=16384
CONFIG_MBEDTLS_SERVER_NAME_INDICATION=y
```

**Comment by joelguittet** [comment]
@jukkar maybe you know is there a good tutorial somewhere to perform debugging ? I mean something official supported by Zephyr. If i go step by step probably I can catch the error when this occurs. I found several vscode information but plenty of them outdated or too much specific, not succeed to use it properly :(

**Comment by joelguittet** [comment, marked as answer]
@jukkar I just increased `CONFIG_MBEDTLS_HEAP_SIZE=100000` (with no real hope this can fix the issue to be honest), and it's now working properly ! i have also set `CONFIG_POSIX_MAX_FDS=16`.

By the way, 100000 looks very large and 60000 was already very large. Any advice to configure this ? I will check in parallel other topics about this setting, I need more documentation about that, if no information found I will open anew discussion.

**Reply by jukkar** [reply]
Agreed, the memory amount sounds excessive. I just recently had similar kind of issues with TLS, the connection establishment was failing. It required me adding prints all over the TLS code path to see where `ENOMEM` was returned. It did not took too long time and in my case the culprit was the `CONFIG_NET_SOCKETS_POLL_MAX=3` which only allowed one TLS connection establishement at a time.
I try to do the debugging in native_sim board as then Zephyr is just a process in my Linux host and it makes investigation easier (at least for networking issues).
